x/incentive/keeper/store.go

```diff
@@ -0,0 +1,274 @@
+// Returns the program's status if found, or an error if it does not exist.
+func (k Keeper) GetIncentiveProgram(ctx sdk.Context, id uint32) (
+	incentive.IncentiveProgram, incentive.ProgramStatus, error,
+) {
+	statuses := []incentive.ProgramStatus{
+		incentive.ProgramStatusUpcoming,
+		incentive.ProgramStatusOngoing,
+		incentive.ProgramStatusCompleted,
+	}
+
+	kvStore := ctx.KVStore(k.storeKey)
+
+	// Looks for an incentive program with the specified ID in upcoming, ongoing, then completed program lists.
+	for _, status := range statuses {
+		program := incentive.IncentiveProgram{}
+		bz := kvStore.Get(keyIncentiveProgram(id, status))
+		if len(bz) != 0 {
+			err := k.cdc.Unmarshal(bz, &program)
+			return program, status, err
+		}
+	}
+
+	return incentive.IncentiveProgram{}, 0, incentive.ErrNoProgramWithID
+}
+
+// setIncentiveProgram stores an incentive program in either the upcoming, ongoing, or completed program lists.
+// does not validate the incentive program struct or the validity of its status change (e.g. upcoming -> complete)
+func (k Keeper) setIncentiveProgram(ctx sdk.Context,
+	program incentive.IncentiveProgram, status incentive.ProgramStatus,
+) error {
+	keys := [][]byte{
+		keyIncentiveProgram(program.Id, incentive.ProgramStatusUpcoming),
+		keyIncentiveProgram(program.Id, incentive.ProgramStatusOngoing),
+		keyIncentiveProgram(program.Id, incentive.ProgramStatusCompleted),
+	}
+
+	kvStore := ctx.KVStore(k.storeKey)
+	for _, key := range keys {
+		// always clear the program from the status it was prevously stored under
+		kvStore.Delete(key)
+	}
+
+	key := keyIncentiveProgram(program.Id, status)
+	bz, err := k.cdc.Marshal(&program)
+	if err != nil {
+		return err
+	}
+	kvStore.Set(key, bz)
+	return nil
+}
+
+// GetNextProgramID gets the ID that will be assigned to the next incentive program passed by governance.
+func (k Keeper) GetNextProgramID(ctx sdk.Context) uint32 {
+	return store.GetUint32(k.KVStore(ctx), keyPrefixNextProgramID, 0, "next program ID")
+}
+
+// setNextProgramID sets the ID that will be assigned to the next incentive program passed by governance.
+func (k Keeper) setNextProgramID(ctx sdk.Context, id uint32) error {
+	prev := k.GetNextProgramID(ctx)
+	if id < prev {
+		return incentive.ErrDecreaseNextProgramID.Wrapf("%s to %s", id, prev)
+	}
+	return store.SetUint32(k.KVStore(ctx), keyPrefixNextProgramID, id, 0, "next program ID")
+}
+
+// GetLastRewardsTime gets the last unix time incentive rewards were computed globally by EndBlocker.
+func (k Keeper) GetLastRewardsTime(ctx sdk.Context) uint64 {
+	return store.GetUint64(k.KVStore(ctx), keyPrefixLastRewardsTime, 0, "last reward time")
+}
+
+// setLastRewardsTime sets the last unix time incentive rewards were computed globally by EndBlocker.
+func (k Keeper) setLastRewardsTime(ctx sdk.Context, time uint64) error {
+	prev := k.GetLastRewardsTime(ctx)
+	if time < prev {
+		return incentive.ErrDecreaseLastRewardTime.Wrapf("%s to %s", time, prev)
+	}
```

Comment:
Small enforcement of invariant here - it will be completely impossible, through this setter, to introduce bugs related to reversing block time.